build.go

```diff
@@ -680,10 +702,77 @@ func Build(opts *BuildArgs) error {
 		}
 	}
 
+	// Revert to original working directory
+	if err := os.Chdir(initialWd); err != nil {
+		return err
+	}
+
 	err = oci.GC(context.Background())
 	if err != nil {
 		fmt.Printf("final OCI GC failed: %v", err)
 	}
 
 	return err
 }
+
+// Build a list of stackerfiles using the same options
+// Note the path in the opts is ignored and changed to by each path in paths
+func BuildMultiple(paths []string, opts *BuildArgs) error {
+
+	// Read all the stacker recipes
+	stackerFiles, err := NewStackerFiles(paths, opts.Substitute)
+	if err != nil {
+		return err
+	}
+
+	// Initialize the DAG
+	dag, err := NewStackerFilesDAG(stackerFiles)
+	if err != nil {
+		return err
+	}
+
+	// In case we don't want to build the images found in includes, filter them out
+	sortedPaths := dag.Sort()
+	var finalPaths []string
+	if !opts.NoIncludes {
+		finalPaths = sortedPaths
```

Comment:
I think here, if len(sortedPaths) != len(paths), we should return errors.Errorf("includes specifically disallowed"), right?

Reply:
Actually that's not the intention.

The reason I added this option is in case someone knows he build the layers in the includes separately, and only wants to build the ones defined in the actual stackerfile sent as parameter.

If we add len() check he won't be able to do that, he'd have to remove/comment the includes altogether.
Do we want this?

Reply:
What do you mean by "built them separately"? Surely we want to rebuild if things have changed according to the cache, right? Or maybe I don't understand.

Reply:
So let's say we have:
- 3 layers, A, B and C. A is base for B, B is base for C.
- A is on one repo. B and C are on another repo. 
So the include will be only in the stacker.yaml of image C, and mention image B.
- The stacker yaml of B has a docker url to image A in 'from'. 
The stacker yaml of C has a docker url to image B in 'from'.

In this scenario 
1. If I run the build for C with the default options it would:
- download C, if not already present in the cache
- run the build for B
- run the build for C, but because of the 'include' directive, the build would ignore the docker url to B, and use the one in the oci layout instead, or ignore the url in the stacker yaml and use the one provided for caching with a git-<commit> tag.
(this last part if not implemented right now, but I was thinking about it)

2. If I run the build for C with the '--no-includes' option it would:
- download B, if not already present in the cache
- run the build for C, ignoring the include directive, and basically behave as if the stacker.yaml did not have it.

So for the same updated stacker.yamls containing includes, 1. would enable us to use the git-commit caching, and 2. would enable us to build same as we do right now without the caching/recursive build feature.

Reply:
Also in the example above, in case 1, even without commit tags. There is no way to avoid building B, if it has not changed. If would be always built together with C.

Reply:
> run the build for C, but because of the 'include' directive, the build would ignore the docker url to B, and use the one in the oci layout instead

I don't think we should do this though. My understanding of include: was just that it would allow you to specify type: built sources that weren't in the current file, but were in one of the files that was included. I don't think we should be re-writing layer sources on the fly, other than via --substitute.

Reply:
My understanding of the include is slightly different, it loosely specifies the order in which the files need to be built. I was wondering if include is really the right name. The name Include would also suggest the content of the stackerfiles is merged into a single list of layers before the build starts (in a single Stackerfile struct instance). But if we go this route the logic for caching stuff in a registry remotely would become even more complicated, and we’d also need to be extra careful with the relative imports.

Configuration names asside, if we’re not re-writing build sources, in order for layer caching to work, we have to upload B to the registry, to the same url specified in the stackerfile for C, before C is built. I guess we could do this by adding configuration for upload url and credentials for B. And use —substitute for the tags.

I would still use the —no-includes to just ignore the includes in the files. My concern is without it there wouldn’t be a way to build just one layer, we’d always have to built the whole tree and upload it to the registry. If we’re ok with rebuilding all layers specified in the includes, and rely on cache to skip the rebuild, then I can make the change.

Reply:
> My understanding of the include is slightly different, it loosely specifies the order in which the files need to be built. I was wondering if include is really the right name. The name Include would also suggest the content of the stackerfiles is merged into a single list of layers before the build starts (in a single Stackerfile struct instance).

Yeah, it's possible that we need a different name. I was thinking that it would behave like an "include" when I suggested it, though, regardless of the actual implementation.

> Configuration names asside, if we’re not re-writing build sources, in order for layer caching to work, we have to upload B to the registry, to the same url specified in the stackerfile for C, before C is built.

I don't think we want to be doing intermediate uploads/downloads at all. What I'm suggesting is that stacker files with includes would look like:

```
/tmp/foo cat a.yaml 
a:
    from:
        type: docker
        url: docker://centos:latest
    run: ...
/tmp/foo cat b.yaml 
stacker_config:
    includes:
        - a.yaml
b:
    from:
        type: built
        tag: a
    run: ...
```

If we do it this way, I'm not sure an option called --no-includes really makes sense; the builds won't work without it. I'm afraid that doing some re-writing on the fly just seems too magical.

I understand that you want to do some build caching in a remote repo somewhere, but I think that's probably config that either lives in ~/.config/stacker.conf or is passed in via the command line, not necessarily a first class thing that users need to reason about when writing stacker files.

Reply:
I understand this model. But in order for incremental builds to work, we need to be able to skip the rebuild of 'a'  in your example, and pick it up from somewhere else. 

In your case let's assume both 'a' and 'b' import sources which need to be built or just copied in the image.

Now let's say for example yesterday I modify the sources imported in the stacker yaml of 'a', this means both 'a' and 'b' (which depends on 'a') needed to be rebuilt, and they were both rebuilt and uploaded to the remote registry to be consumed by various users.

Tomorrow, someone else, in a different build environment, modifies the sources imported in 'b'. 'a' does not have to be rebuilt since the imports have not changed since yesterday, and he should only have to wait for 'b' to be built. Yesterday's version of 'a' would be downloaded, and used as base for 'b'. 

So the build for 'b' needs to account both for the case when 'a' needs to be rebuilt, and the case in which 'a' doesn't need to be rebuilt. Always pushing 'a' to a repository, and optionally downloading it in case we need to skip it's build are part of the solution.

This is the kind of decision making we're currently doing in our Jenkins groovy + bash code. This is a real use case. 
And we need unify the build logic, so the build behaves the same way regardless if we build it in Jenkins or the developer does it on his individual machine.

If stacker is not the right place to make these kind of decisions, let's figure out which decisions should be made by stacker, and which should be done by other tools, be they in bash or go.

Reply:
> > My understanding of the include is slightly different, it loosely specifies the order in which the files need to be built. I was wondering if include is really the right name. The name Include would also suggest the content of the stackerfiles is merged into a single list of layers before the build starts (in a single Stackerfile struct instance).
> 
> Yeah, it's possible that we need a different name. I was thinking that it would behave like an "include" when I suggested it, though, regardless of the actual implementation.
> 
> > Configuration names asside, if we’re not re-writing build sources, in order for layer caching to work, we have to upload B to the registry, to the same url specified in the stackerfile for C, before C is built.
> 
> I don't think we want to be doing intermediate uploads/downloads at all. What I'm suggesting is that stacker files with includes would look like:
> 
> ```
> /tmp/foo cat a.yaml 
> a:
>     from:
>         type: docker
>         url: docker://centos:latest
>     run: ...
> /tmp/foo cat b.yaml 
> stacker_config:
>     includes:
>         - a.yaml
> b:
>     from:
>         type: built
>         tag: a
>     run: ...
> ```

The 'tag: a" should probably be "tag: a:a" ?  Else we're placing a new demand that all image builds in a repo have unique names.

Reply:
> Now let's say for example yesterday I modify the sources imported in the stacker yaml of 'a', this means both 'a' and 'b' (which depends on 'a') needed to be rebuilt, and they were both rebuilt and uploaded to the remote registry to be consumed by various users.

> Tomorrow, someone else, in a different build environment, modifies the sources imported in 'b'. 'a' does not have to be rebuilt since the imports have not changed since yesterday, and he should only have to wait for 'b' to be built. Yesterday's version of 'a' would be downloaded, and used as base for 'b'.

I think we can do this in stacker, but I don't think this should be expected to be the default behavior. When you're first starting out with stacker, you're not going to have a setup as complex as we do, and you'll want to do includes, etc. just to play around with it, without necessarily uploading it. I think the default way to express things in the yaml should reflect this.

Now, we can definitely do something where we either accept via arguments, additional config in the build yaml, etc. a URL where potential builds might be cached (with a naming scheme dictated by stacker) so that stacker can try to resolve things if they've already been built. So I think we can still accomplish what you want, with additional configuration to make explicit what's happening, instead of implicitly rewriting stuff.

Reply:
In this case I'll just proceed with addressing the original comments in this PR. 
Will also remove the --no-includes directive.
And add tests.

Maybe change the name from 'includes' to 'prerequisites'.

Reply:
Tycho,

Made some progress, but having some difficulty with how stacker handles the layers:
https://github.com/anuvu/stacker/pull/35/commits/f78ac09ac48e05920d59f1bf8cf1fea24ef3196b

Stacker assumes all layers are defined in the same stackerfile, and actively searches the stackerfile for the base layers. I've added some warnings instead of errors in a couple of places, but I don't know how to handles the one below.

It's the case when a layer depends on another layer of type  'BuiltType'.

Partial output from the new tests:
`# building image layer2...
# importing files...
# copying import2
# WARNING: base layer layer1_1 mising from stackerfile /home/andaaron/anuvu/stacker/ocibuilds/sub2/stacker.yaml!
# WARNING: base layer layer1_1 mising from stackerfile /home/andaaron/anuvu/stacker/ocibuilds/sub2/stacker.yaml!
# running commands...
# running commands for layer2
# + cp /stacker/import2 /root/import2
# + cp /root/import1 /root/import1_copied
# generating layer for layer2
# setting git version annotation to 9b158a7a4545c7dc022542cdb661acc746cbd984-dirty
# filesystem layer2 built successfully
# error: couldn't find a cache of base layer
`
Looks like it tries to look for the base layer in the cache, after the build, but it doesn't find it. The cache is coupled to the stackerfile, so not sure how to go about this.

Reply:
> The cache is coupled to the stackerfile, so not sure how to go about this.

I think that's the bug: in your test, you have sub1, sub2, sub3, but you only call the actual "stacker build" in sub3; stacker shouldn't write any files in any other directory besides sub3 in this case.

Reply:
I'm actually calling stacker outside those folders and sending the file as parameter with '-f'.

Is there an assumption the cache is always located relative to the current working directory?
In this case the bug should go away after I get rid of chdir.

Reply:
> I'm actually calling stacker outside those folders and sending the file as parameter with '-f'.
Seems like you probably want to pass --stacker-dir then, to make it whatever the original user passed as --stacker-dir (same for --oci-dir and --roots-dir, then). But it might be nice to try and do this all in golang, rather than fork+exec.

> Is there an assumption the cache is always located relative to the current working directory?
No, the cache is always located relative to --stacker-dir though.

Reply:
This is not the issue, the code already accounts for the --stacker-dir as an absolute path.

The issue seems to be at:
https://github.com/anuvu/stacker/blob/master/build.go#L679
which leads to
https://github.com/anuvu/stacker/blob/master/cache.go#L300
which leads to
https://github.com/anuvu/stacker/blob/master/cache.go#L281
which leads to
https://github.com/anuvu/stacker/blob/master/cache.go#L162

It's looking for a layer with the name of the base layer in the same stackerfile.

Reply:
Ah, ok. Seems like that resolution will have to be updated then too.

Reply:
How to take care of it? If it was straight forward I wouldn't have asked :)
Also do you think it would make more sense to switch back to the name 'include', concatenate all layers in the same Stackerfile object, and call build() only once? Would that make more sense?

Reply:
> How to take care of it?

I think the reason for needing the stackerfiles in the cache is so it can look up each individual layer, to figure out if the structure has changed (e.g. new run section, import, or bind mount). So I think you can just make OpenCache() take a []*Stackerfile instead, and it should just look through all of them sequentially.

> Also do you think it would make more sense to switch back to the name 'include', concatenate all layers in the same Stackerfile object, and call build() only once? Would that make more sense?

Oh, yeah, that may make things super easy. Just open them all, cat them all together, run substitutions, and parse the result? I guess you might have to do something slightly fancy with the stacker_config: section, but other than that it seems like maybe that would simplify all of this?